### ISSUE TITLE
Signed packages

### ISSUE THREAD

**Issue by ViralBShah** [JuliaLang/julia#1819]
We should have the ability to have GPG signed packages.

**Comment by mauro3** [comment]
This is being discussed and advanced here https://groups.google.com/d/msg/julia-dev/T246S9KyMJY/sdUJBg5IBgAJ by @LachlanGunn (I think that is his github name).

**Comment by LachlanGunn** [comment]
Yes, that's me.

**Comment by wildart** [comment]
But we already have signing process for packages. Every package has a version which is a tagged commit that is identified by a SHA1 hash. This hash is stored in METADATA. When package is installed a particular commit, that corresponds to the stored SHA1 hash, is checked out. And so we get a verified version of code.

**Comment by tkelman** [comment]
sha1 isn't cryptographically secure any more, but I don't think libgit2 supports gpg signing so if we're going to do this somehow I believe it would require going through command-line git.

**Comment by LachlanGunn** [comment]
The other reason I'm not a big fan of signed commits is that it really strongly ties you to git, I would be keen to be able to validate just a directory without any history, or maybe data as well. 

I've been kind on silent on this for a while, but I've needed to write a partial PGP implementation for something, so maybe that could be useful for the purposes of avoiding GPL and LGPL. It uses MbedTLS for the actual crypto, but I'd be hesitant to suggest it without some auditing.

**Comment by StefanKarpinski** [comment]
I agree: we should use a more recent cryptographically secure hash (BLAKE2b-512 and/or SHA-3-512), and we should only hash the file tree, not the commit. Since we track the git commit SHA1 as well, which includes all history, it will be impossible to tamper with the content or the history, although content is most important. I would advocate using [libsodium](https://download.libsodium.org/doc/) for our crypto – it is ISC-licensed (even more liberal than MIT) and provides all the primitives we would need for hashing and eventually signing.

**Comment by LachlanGunn** [comment]
While I used libsodium (or a compatible library, anyway)  in my prototype, I'm not sure it's ideal for our needs because it doesn't include any key management infrastructure, and can't AFAIK be used with smartcards.

It's a good library if you're trying to do something from scratch, but do we need to do this from scratch given how many other systems have already done it?

**Comment by wildart** [comment]
MbedTLS has SHA-512.

P.S. https://mikegerwitz.com/papers/git-horror-story

**Comment by staticfloat** [comment]
Note that we do have [native hashing in Julia](https://github.com/staticfloat/SHA.jl), so we don't need to include `libsodium` if we're okay with putting some more code into `Pkg`.

**Comment by StefanKarpinski** [comment]
For version integrity checking, native Julia would be ok.

**Comment by LachlanGunn** [comment]
Yep, for that it's fine, but libsodium/NaCL/etc. all provide implementations of the elliptic curve stuff, which is nontrivial to implement and best not to touch if at all possible---even Bernstein's code had an error in it that wasn't found for some time IIRC.

**Comment by ViralBShah** [comment]
I believe the Pkg3 design allows for this. Should we close this?

**Comment by KristofferC** [comment]
Not until Pkg3 is released and usable?

**Comment by StefanKarpinski** [comment]
Signing is not part of the core Pkg3 design. Support for better hash algorithms is, which should thoroughly address integrity checking issues. Signing of packages with private keys can be done through opt-in metadata in registries, but pretty careful thought needs to be put into how that works in order for it not to be security theater.

**Comment by StefanKarpinski** [comment]
In any case, I don't think the language itself needs to do anything to support signed packages.
